{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Programming for Biomedical Informatics\n",
    "#### Week 11 - Revision Session 2 - Psuedocoding Examples\n",
    "\n",
    "We're going to go through a few different examples of psuedocoding to show you what we are looking for in exam answers for structure, granularity, and commenting.\n",
    "\n",
    "The principle we are applying to guide this is that we expect a suitably experienced researcher to be able to take the pseudocoding prompts you include in your answer and write the specfic lines of corresponding code. Writing a high-level pseudocode line that would need to be coded by multiple lines of code is not ordinarily sufficient. For example, if you were writing a `for` loop to perform a caclulation over elements of a list it would not be sufficient to simply write that in one line, you would be expected to indent smaller, more specific, pseduocode instructions on what would become each line of code in the final coded implementation. For a `for` loop it could be that you intend the coder to use a list comprehension which can be implemented in one line of code. In that case an explanation of how the list comprehension needs to be constructed should be included in the psuedocode comment as it's syntax and formulation is harder to write than a step by step process implemented in multiple lines of a `for` loop.\n",
    "\n",
    "Below we present a number of examples. We will use a reverse approach, starting with a complete commented code snippet and deconstructing it back down to pseduocode.\n",
    "\n",
    "We recommend that you practice this by taking some of your own code or code from the various notebooks in the course and going through the same process.\n",
    "\n",
    "In the following examples we will use:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' to contain explanatory text and '''\n",
    "\n",
    "# at the start of lines that contain the actual pseudocode text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 1 - Original Code - Single Lines\n",
    "\n",
    "'''This very simple example has no complex structure, simply single lines of code. The key here is to make sure that the commenting and the pseduocode are clear and concise.\n",
    "The code is simple enough that the pseudocode is not really necessary, but it is included for demonstration purposes.'''\n",
    "\n",
    "'''This code reads in a gene coexpression network and a gene expression dataset, and prints the number of nodes in the gene coexpression network'''\n",
    "\n",
    "# import the required libraries\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "\n",
    "#Â load the gene coexpression network\n",
    "\n",
    "# define the data directory\n",
    "data_dir = './data/'\n",
    "\n",
    "# Define paths to .gml network files from Section 1\n",
    "# These paths point to the Gene correlation network, Patient network from (1) TCGA gene expression data and (2) TCGA DNA methylation data.\n",
    "G_gxp_path = data_dir + 'gene_coexpression_network.gml'\n",
    "\n",
    "# Load the GML graphs into NetworkX graph objects\n",
    "# nx.read_gml() function reads a graph from a GML file\n",
    "G_gxp = nx.read_gml(G_gxp_path)  # Gene correlation network\n",
    "\n",
    "# Get all nodes in each graph\n",
    "# The nodes represent genes or patients depending on the network\n",
    "G_gxp_nodes_list = list(G_gxp.nodes())  # Nodes in the gene correlation network\n",
    "\n",
    "# Define paths to the raw TCGA datasets\n",
    "# tcga_dnam_path = 'section2_data/ISMB_TCGA_DNAm.pkl'  # TCGA DNA methylation data\n",
    "tcga_gxp_path = data_dir + 'ISMB_TCGA_GE.pkl'  # TCGA Gene expression data\n",
    "\n",
    "# Load the gene expression dataset\n",
    "# pd.read_pickle() function loads a pickled pandas DataFrame or Series\n",
    "tcga_gxp = pd.read_pickle(tcga_gxp_path)\n",
    "\n",
    "# For this example, we'll use a CSV file that includes gene symbols\n",
    "# pd.read_csv() function loads a CSV file into a pandas DataFrame\n",
    "tcga_gxp_df = pd.read_csv(data_dir + 'tcga_ge_df_symbols_t.csv') # Dataset with gene symbols\n",
    "# Set 'GENES' column as the index for easy access to gene-specific data\n",
    "tcga_gxp_df.set_index('GENES', inplace=True)\n",
    "\n",
    "# Extract metadata from the gene expression dataset\n",
    "# Metadata might include information such as patient IDs, sample conditions, etc.\n",
    "tcga_gxp_meta = tcga_gxp['datMeta']\n",
    "\n",
    "# Print the number of nodes in each network\n",
    "# This provides a quick overview of the size of each network\n",
    "print(f\"Number of nodes in gene correlation network: {len(G_gxp_nodes_list)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 1 - Pseudocode\n",
    "\n",
    "'''\n",
    "In this example I have broadly generalised the specific code above, but it would also have been fine to use specific parts of the code in the pseudocode and to explictly specify the names of files and variables.\n",
    "The latter can be especially useful if you will later perform operations on those variables, as it can help you to keep track of what you are doing. In this case it's implicit which variable we will use because it's\n",
    "a simple example, but in more complex examples it can be useful to be explicit.\n",
    "\n",
    "It's good practice to use a explanatory comment at the start of the pseudocode to explain what the code does, and to use comments to explain the intended function of each line of code. This is the approach we want you to use in the exam.\n",
    "'''\n",
    "\n",
    "# PSEUDOCODE START\n",
    "\n",
    "'''\n",
    "This code reads in a gene coexpression network, gene expression dataset, and an associated meta-data file and prints the number of nodes in the gene coexpression network. It specifies the loading of a series of files that \n",
    "will be used later in a downstream analysis pipeline.\n",
    "\n",
    "For this code we will need three data files:\n",
    "- A gene coexpression network in GML format\n",
    "- A gene expression dataset in pickle format\n",
    "- A metadata file in CSV format\n",
    "\n",
    "pandas will be needed to read in meta-data from a pickle file\n",
    "networkx will be needed to read in the gene coexpression network from a gml format file\n",
    "'''\n",
    "\n",
    "# import the pandas and networkx libraries\n",
    "\n",
    "# set a variable to store the relative file path of the base directory where data is located\n",
    "\n",
    "# define two variables to store the file paths of the gene coexpression network and the gene expression data, respectively\n",
    "\n",
    "# Load the GML file for the gene expression network into NetworkX using the read_gml function\n",
    "\n",
    "# Get a list of all the nodes in the gene expression network using the nodes() function of networkx and store them in a variable\n",
    "\n",
    "# Load the gene expression dataset from the pickle file using the read_pickle function of pandas and store in a variable\n",
    "\n",
    "# Load the meta-data file that contains gene symbols from a CSV file using the read_csv function of pandas and store in a variable\n",
    "\n",
    "# Re-index the dataframe so that the 'GENES' column is the index for easy access to gene-specific data using the set_index function of pandas\n",
    "\n",
    "# Extract metadata from the gene expression dataset using the 'datMeta' key and store in a variable\n",
    "\n",
    "# Print the number of nodes in each network\n",
    "\n",
    "\n",
    "# PSEUDOCODE END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 2 - Simple nested code in a function\n",
    "\n",
    "# the human phenotype ontology contains database_cross_reference entries that include the UMLS concept id\n",
    "# we can use this to link the HPO terms to the UMLS concepts\n",
    "\n",
    "# load the HPO data using pronto\n",
    "import pronto\n",
    "\n",
    "# load the HPO ontology\n",
    "# fetch the Human Phenotype Onology OBO file and parse it with pronto\n",
    "\n",
    "# download the HPO ontology OBO file\n",
    "import urllib.request\n",
    "\n",
    "current_hpo_url = 'http://purl.obolibrary.org/obo/hp.obo'\n",
    "\n",
    "# download the file\n",
    "urllib.request.urlretrieve(current_hpo_url,'hpo.obo');\n",
    "\n",
    "# parse the file\n",
    "hpo = pronto.Ontology('hpo.obo')\n",
    "\n",
    "# we can look in the xrefs (sic. cross-references) of a term to find the UMLS concept id\n",
    "def hpo2concept(hpo_id):\n",
    "    term = hpo[hpo_id]\n",
    "    xrefs = [xref.id for xref in term.xrefs]\n",
    "    try:\n",
    "        umls_id = [xref for xref in xrefs if xref.startswith('UMLS')][0].split(':')[1]\n",
    "        return umls_id\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "# let's test this function\n",
    "print(hpo2concept)('HP:0001695')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 2 - Pseudocode\n",
    "\n",
    "'''\n",
    "Note the indentations!\n",
    "These should be used in loops the same way that they would in actual code.\n",
    "'''\n",
    "\n",
    "# PSEUDOCODE START\n",
    "\n",
    "'''\n",
    "This code reads in the Human Phenotype Ontology (HPO) and defines a function that takes an HPO term ID as input and returns the corresponding UMLS concept ID.\n",
    "\n",
    "pronto will be needed to read in the Human Phenotype Ontology (HPO) data\n",
    "urllib will be needed to download the HPO ontology file\n",
    "'''\n",
    "\n",
    "# load the relevant libraries\n",
    "\n",
    "# store the URL for the HPO OBO file in a variable\n",
    "\n",
    "# use urllib.request to download the HPO OBO file\n",
    "\n",
    "# use pronto to load the onology from the OBO file\n",
    "\n",
    "'''\n",
    "Define a function that will take an accession ID from the Human Phenotype Ontology (HPO) as input and return any cross-references from the Unified Medical Language System (UMLS).\n",
    "\n",
    "Note: the function should use exception handling to return None if no cross-references are found.\n",
    "\n",
    "To achieve this we will need to use:\n",
    "\n",
    "- An OBO format file for the HPO ontology, which can be downloaded from the NCBO BioPortal\n",
    "- The urllib.request library to download the HPO ontology file using the corresponding URL found on the NCBO BioPortal website\n",
    "- The `pronto` library to parse the OBO file\n",
    "'''\n",
    "\n",
    "# Define a function called hpo2concept that takes a single argument, hpo_id\n",
    "\n",
    "# def hpoconcept(hpo_id)\n",
    "    # Retrieve the term corresponding to the hpo_id from the hpo dictionary\n",
    "\n",
    "    # Extract the cross-references for the term and store them in a list called xrefs\n",
    "\n",
    "    # open a try: except: block to handle exceptions\n",
    "\n",
    "    # try:\n",
    "        # Use a list comprehension to retrieve the cross-references fot the provided ontology term from the xrefs list\n",
    "            # The cross-references are stored as strings in a CURIE format 'database:accession', e.g., 'UMLS:C0007222'\n",
    "            # Extract the database and accession separately to check for a UMLS cross-reference and extract the UMLS concept ID\n",
    "        # return the UMLS concept IDs\n",
    "\n",
    "    # except:\n",
    "        # return None\n",
    "\n",
    "# Test the function by calling it with a specific HPO ID that you know from checking on the NCBO BioPortal website has a UMLS cross-reference associated with it and print the result\n",
    "\n",
    "# PSEUDOCODE END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 3 - More complex nested code in a function\n",
    "\n",
    "# the ASDPTO ontology is a custom ontology for autism spectrum disorder the URL for the ontologu file is:\n",
    "current_asdpto_url = 'https://data.bioontology.org/ontologies/ASDPTO/submissions/1/download?apikey=4a2fbff0-ef88-432e-b1a1-dffc07e71146'\n",
    "\n",
    "# download the file\n",
    "urllib.request.urlretrieve(current_asdpto_url,'autism.obo');\n",
    "\n",
    "# parse the file using pronto\n",
    "autism  = pronto.Ontology('autism.obo')\n",
    "\n",
    "# function to find the UMLS concept for a term in ASDPTO\n",
    "def find_concept(term):\n",
    "    for annotation in term.annotations:\n",
    "        try:\n",
    "            # if the string contains a cui= then it is a UMLS concept\n",
    "            # extract the CUI\n",
    "            if 'cui=' in annotation.resource:\n",
    "                #split the string on 'cui=' and take the remainder\n",
    "                concept = annotation.resource.split('cui=')[1]\n",
    "                return(concept)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "# we can now use this function to find the UMLS concept for each term in the ASDPTO ontology\n",
    "asdpto2umls = {term.name:find_concept(term) for term in autism.terms()}\n",
    "\n",
    "# remove any None entries\n",
    "asdpto2umls = {k:v for k,v in asdpto2umls.items() if v is not None}\n",
    "\n",
    "# print how many terms have been mapped to UMLS concepts\n",
    "print(f'There are ',len(asdpto2umls),' terms in the ASDPTO ontology that have been mapped to UMLS concepts')\n",
    "\n",
    "# look at the first 10 entries\n",
    "list(asdpto2umls.items())[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Example 3 - Pseudocode\n",
    "\n",
    "'''\n",
    "NB Note the indentations! especially the use of if: else: and try: except: blocks which are indentations that should be used in loops the same way that they would in actual code.\n",
    "'''\n",
    "\n",
    "# PSEUDOCODE START\n",
    "'''\n",
    "This code reads in the Autism Spectrum Disorder Phenotype Ontology (ASDPTO) and defines a function that takes an ASDPTO term as input and returns the corresponding UMLS concept ID.\n",
    "\n",
    "pronto will be needed to read in the Autism Spectrum Disorder Phenotype Ontology (ASDPTO) data\n",
    "urllib will be needed to download the ASDPTO ontology file\n",
    "'''\n",
    "\n",
    "# load the relevant libraries\n",
    "\n",
    "# store the URL for the ASDPTO OBO file in a variable\n",
    "\n",
    "# use urllib.request to download the ASDPTO OBO file\n",
    "\n",
    "# use pronto to load the onology from the OBO file\n",
    "\n",
    "'''\n",
    "Define a function that will take an accession ID from the Autism Spectrum Disorder Phenotype Ontology (ASDPTO) as input and return any cross-references from the Unified Medical Language System (UMLS).\n",
    "\n",
    "Note: the function should use exception handling to return None if no cross-references are found.\n",
    "\n",
    "To achieve this we will need to use:\n",
    "\n",
    "- An OBO format file for the HPO ontology, which can be downloaded from the NCBO BioPortal\n",
    "- The urllib.request library to download the HPO ontology file using the corresponding URL found on the NCBO BioPortal website\n",
    "- The `pronto` library to parse the OBO file\n",
    "'''\n",
    "\n",
    "# Define a function called find_concept that takes a single argument, a term from the ASDPTO ontology\n",
    "\n",
    "# function to find the UMLS concept for a term in ASDPTO\n",
    "\n",
    "# def find_concept(term):\n",
    "    # for each annotation that the term object has:\n",
    "        #try:\n",
    "            # check whether the resource slot of the annotation object is a UMLS concept\n",
    "            # if the string contains `cui=``:\n",
    "                # split the string on 'cui=' and take the remainder as the UMLS concept ID\n",
    "                # return(concept)\n",
    "            # else:\n",
    "                # pass\n",
    "        # except:\n",
    "            # pass\n",
    "\n",
    "# use this function to find the UMLS concept for every term in the ASDPTO ontology and store the results in a dictionary\n",
    "# the dictionary should have the term name as the key and the UMLS concept ID as the value\n",
    "# use a lists comprehension to iterate over the terms in the ASDPTO ontology and call the find_concept function on each term\n",
    "\n",
    "# check and remove any `None` (empty) entries from the dictionary using a dictionary comprehension\n",
    "\n",
    "# print how many terms have been mapped to UMLS concepts using a printf statement\n",
    "\n",
    "# print the first 10 entries in the dictionary"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "networks",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
